### ISSUE TITLE
Feature request: ngBeforeDestroy

### ISSUE THREAD

**Issue by JurajMlich** [angular/angular#17404]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When ngOnDestroy is called, the element is already removed from the DOM. 

**Expected behavior**
Before ngOnDestroy and before removing from DOM, ngBeforeDestroy would get called.

**What is the motivation / use case for changing the behavior?**
I have a list of items. When the currently focused one is removed, I want to focus the next item. However, due to the element already being out of DOM, I do not know what the next item in DOM is.

Thank you.

**Comment by dustincleveland** [comment]
This would be really helpful as it would address #14252. Is there any way to achieve this functionality in the short term?

**Comment by glebmachine** [comment]
Any updates?

**Comment by dasAnderl** [comment]
this is a must. working with tradingview charts. ngOnDestroy is called after the dom (in my case the chart container) is removed. any call to chart js object results then in parenNode not found. this is just one example.
also the name implies that it executes before dom is removed

**Comment by dustincleveland** [comment]
In my use case, I believe I was able to find a successful work around to clean up my third-party event handlers.

TS:
```
@ViewChild('chart') public chart: ElementRef;
...
Plotly.purge(this.chart.nativeElement);
```

HTML:
```
<div #chart
    class="chart"
    [id]="chartId">
</div>
```

The elementRef isn't disposed yet, which seems to allow me to successfully reference the DOM element I need.

**Comment by radoslavpetranov** [comment]
Having a hook that is being called right before the element is about to get destroyed but while it's still visible on screen would be really useful. It's very easy to make simple transitions using rxjs in `ngOnInit`

```
ngOnInit() {
    observableOf(true).pipe(
      take(1),
      tap(val => this.isRendered = true),
      delay(50),
      tap(val => this.isShown = true),
    ).subscribe();
  }
```
where `this.isRendered` is just a hostListener that sets the display from `none` to `flex` and `this.isShown` is the class that takes care of the transition.

One could then use `ngBeforeDestroy` in a similar fashion to update the HostListeners accordingly and exit the component cleanly.

Obviously this can be achieved in other ways but the other approaches require more code, more logic and more rxjs to compensate for a missing lifecycle hook. This is not a show stopper but definitely a useful addition to the existing hooks.

**Comment by glebmachine** [comment]
That's looks weird and this is are weird, but, in other hand, it's a solution) 🍺

@radoslavpetranov, could you please share full implementation of component, works as you described? That will be appreciated by struggling muggles)

For example, you can use https://stackblitz.com/ for ease sharing case of use

**Comment by simeyla** [comment]
I have created a `NgViewRefOutlet` directive, which works like `NgTemplateOutlet` except with ViewRefs. The idea is so I can move DOM elements around on the page easily without losing state. However I cannot detach a ViewRef inside `ngOnDestroy` because it's already destroyed.

**Comment by Goldflow** [comment]
I agree this feature should be there.  Not sure if it will be of help to anyone, but we are handling our navigation in a wrapper service (because apart from routing with Router, for the other half of pages we route to a page containing an iframe that points to the requested iframe url).

Upon calling the function to handle a navigaton request, I'm triggering a navigationIntent$ subject which will then inform any subscribers when the user is about to navigate to another page, effectively giving me a hook before onDestroy is called... but it's only useful in certain scenario's.

**Comment by Wagan8r** [comment]
This is exactly what my team and I need to properly clean up third party frameworks when the host (Angular) component is destroyed. We have a React component that needs to be unmounted, and we have no hook to do so as it is already wiped from the DOM by the time it's wrapper Angular component has been destroyed.

**Comment by edusperoni** [comment]
I'm currently working on some changes on the recycling of views for the NativeScript framework, and turns out I also need this.

I dug around in the code and I think an additional function needs to be created and called before line 283:

https://github.com/angular/angular/blob/7ca611cd1266ede26a26b17cb31748988ff8d98c/packages/core/src/render3/node_manipulation.ts#L281-L290

From my understanding, the renderer destroy all nodes in `executeActionOnView(renderer, WalkTNodeTreeAction.Destroy, lView, null, null);`, then it goes down the tree calling `OnDestroy`:

https://github.com/angular/angular/blob/7ca611cd1266ede26a26b17cb31748988ff8d98c/packages/core/src/render3/node_manipulation.ts#L338

So we'd need to either:

1. Go down the tree executing `ngOnBeforeDestroy` before the view is destroyed by the renderer.
2. Or move this code to the `executeActionOnView` code and call the event before:
https://github.com/angular/angular/blob/7ca611cd1266ede26a26b17cb31748988ff8d98c/packages/core/src/render3/node_manipulation.ts#L94-L97

**Comment by edusperoni** [comment]
I've created a sample repo for this: https://github.com/edusperoni/ngbeforedestroy

I've also tested my proposed implementation, and I'm able to manipulate my `ViewContainerRef` before it's destroyed.

Proof of concept:

Edit the file `node_modules/@angular/core/fesm2015/core.js`. look for this code and add the marked line:
```typescript
function destroyView(view) {
    if (view.state & 128 /* Destroyed */) {
        return;
    }
    callLifecycleHooksChildrenFirst(view, 131072 /* OnDestroy */); // ADD THIS LINE
    execEmbeddedViewsAction(view, ViewAction.Destroy);
    execComponentViewsAction(view, ViewAction.Destroy);
    callLifecycleHooksChildrenFirst(view, 131072 /* OnDestroy */);
```

**this proof of concept will call ondestroy twice (I've used the same event, can be remapped)**

Normal operation when the button is clicked and the view is destroyed:

```
toggle ondestroy
inner destroyed
```

New behavior:
```
toggle ondestroy // this is the first destroy, our ngBeforeDestroy. Here I detach inner
toggle ondestroy // second destroy hook, vcref is empty and the views were destroyed by this point
// inner is not destroyed (was detached)
```

**Comment by icholy** [comment]
I have an openlayers component and I need to call `Map.dispose` before the element is removed.

**Comment by myrddraall** [comment]
@simeyla 
Doing something similar
I need the ability do detach ViewRefs before they are marked as destroyed. 
currently if you detach in ngOnDestroy they are already destroyed and attempting to reattach elsewhere results in
`Error: Cannot insert a destroyed View in a ViewContainer!`

**Comment by nishantkmr** [comment]
Any update on this feature request? This has so many up votes and is lingering for such a long time.

**Comment by eneajaho** [comment]
Instead of a new lifecycle hook, can we have this new hook already in DestroyRef class? 

```ts
inject(DestroyRef).onBeforeDestroy(() => {
  // cleanup animations ? 
});
```

**Comment by Char2sGu** [comment]
When a leaving animation is defined, ngOnDestroy is actually invoked BEFORE the element is removed from DOM, which raises another issue https://github.com/angular/angular/issues/48224

**Comment by ivanpadavan** [comment]
I've never known that ngOnDestroy is called after DOM node removal. Actually under my uscases it's always called before node removal and it is essential.

But i unit tests it's invoked after node removal. In one with pending periodic task which makes test behaviour timer-dependant (while in real environment it is not). But in the other I stucked with element being renoved before OnDestroy hook

**Comment by mprakhar1705** [comment]
I am working on a project where in the route doesn't necessarily changes when moving from one component to other (tabs/menus switch etc within the parent component). If a component is dirty and user is navigating to other component, I need to display "You have unsaved changes" confirmation dialog giving user the option to stay or leave.
With the presence of lifecycle hook like _ngBeforeDestroy_ it would become much easier as I can call the dialog service in _ngBeforeDestroy_  without the need to intercept all the possible events.

**Comment by JeanMeche** [comment]
@mprakhar1705 That looks like something should be handled by routing +  a `canDeactivate` guard.

**Comment by mprakhar1705** [comment]
> [@mprakhar1705](https://github.com/mprakhar1705) That looks like something should be handled by routing + a `canDeactivate` guard.

Yes, totally agreed, need to confirm with my team if we can have a different route for every component. Then `canDeactivate` would solve the problem. But as per the current structure, `canDeactivate` won't work and I am intercepting the event to achieve the said requirement. Thanks for the quick reply though!

**Comment by edusperoni** [comment]
Although not intended for it, the new angular leave animations will likely execute before destroy, then destroy the component and animate the leave with the component destroyed. I haven't tested it though
